merlin-server/constraints-dsl-compiler/src/libs/constraints-edsl-fluent-api.ts

```diff
@@ -1,87 +1,324 @@
 import * as AST from './constraints-ast.js';
 
-interface ViolationsOf extends Constraint {}
-
 export class Constraint {
   /** Internal AST node */
   public readonly __astNode: AST.Constraint;
 
-  private constructor(astNode: AST.Constraint) {
+  public constructor(astNode: AST.Constraint) {
```

Comment:
I left a similar comment for Ryan at some point.

The pattern of having a private constructor and a public new function is a bit of an interesting one.

 A constructor MUST be synchronous and must return the class type. A static new function can be asynchronous and does not need to return the class instance. This is useful for cases where you may need an async call as part of the initialization (time conversion via a service is the straightforward example). Or you may want to have a fallible object creation and return a Result type.

As a result of running into these issues requiring extensive refactoring that is less obvious a number of times, I've gotten into the habit of defaulting to the static new pattern.

I'm not saying this pattern is a necessary part of this or any codebase, just some food for thought on the matter.